pkg/util/converter/prom.go

```diff
@@ -0,0 +1,410 @@
+func ReadPrometheusStyleResult(iter *jsoniter.Iterator) *backend.DataResponse {
+	var rsp *backend.DataResponse
+	status := "unknown"
+
+	for l1Field := iter.ReadObject(); l1Field != ""; l1Field = iter.ReadObject() {
+		switch l1Field {
+		case "status":
+			status = iter.ReadString()
+
+		case "data":
+			rsp = readPrometheusData(iter)
+
+		// case "error":
+		// case "errorType":
+		// case "warnings":
+		default:
+			v := iter.Read()
+			fmt.Printf("[ROOT] TODO, support key: %s / %v\n", l1Field, v)
+		}
+	}
+
+	if status != "success" {
+		fmt.Printf("ERROR: %s\n", status)
+	}
+
+	return rsp
+}
+
+func readPrometheusData(iter *jsoniter.Iterator) *backend.DataResponse {
+	t := iter.WhatIsNext()
+	if t == jsoniter.ArrayValue {
+		return readArrayData(iter)
+	}
+
+	if t != jsoniter.ObjectValue {
+		return &backend.DataResponse{
+			Error: fmt.Errorf("expected object type"),
+		}
+	}
+
+	resultType := ""
+	var rsp *backend.DataResponse
+
+	for l1Field := iter.ReadObject(); l1Field != ""; l1Field = iter.ReadObject() {
+		switch l1Field {
+		case "resultType":
+			resultType = iter.ReadString()
+
+		case "result":
+			switch resultType {
+			case "matrix":
+				rsp = readMatrixOrVector(iter)
+			case "vector":
+				rsp = readMatrixOrVector(iter)
+			case "streams":
+				rsp = readStream(iter)
+			default:
+				iter.Skip()
+				rsp = &backend.DataResponse{
+					Error: fmt.Errorf("unknown result type: %s", resultType),
+				}
+			}
+
+		case "stats":
+			v := iter.Read()
+			if len(rsp.Frames) > 0 {
+				meta := rsp.Frames[0].Meta
+				if meta == nil {
+					meta = &data.FrameMeta{}
+					rsp.Frames[0].Meta = meta
+				}
+				meta.Custom = map[string]interface{}{
+					"stats": v,
+				}
+			}
+
+			fmt.Printf("[data] TODO, support stats: %v\n", v)
+
+		default:
+			v := iter.Read()
+			fmt.Printf("[data] TODO, support key: %s / %v\n", l1Field, v)
+		}
+	}
+
+	fmt.Printf("result: %s\n", resultType)
+	return rsp
+}
+
+// will always return strings for now
+func readArrayData(iter *jsoniter.Iterator) *backend.DataResponse {
+	lookup := make(map[string]*data.Field)
+
+	rsp := &backend.DataResponse{}
+	stringField := data.NewFieldFromFieldType(data.FieldTypeString, 0)
+	stringField.Name = "Value"
+	for iter.ReadArray() {
+		switch iter.WhatIsNext() {
+		case jsoniter.StringValue:
+			stringField.Append(iter.ReadString())
+
+		// Either label or exemplars
+		case jsoniter.ObjectValue:
+			f, pairs := readLabelsOrExemplars(iter)
+			if f != nil {
+				rsp.Frames = append(rsp.Frames, f)
+			} else if pairs != nil {
+				max := 0
+				for _, pair := range pairs {
+					k := pair[0]
+					v := pair[1]
+					f, ok := lookup[k]
+					if !ok {
+						f = data.NewFieldFromFieldType(data.FieldTypeString, 0)
+						f.Name = k
+						lookup[k] = f
+
+						if len(rsp.Frames) == 0 {
+							rsp.Frames = append(rsp.Frames, data.NewFrame(""))
+						}
+						rsp.Frames[0].Fields = append(rsp.Frames[0].Fields, f)
```

Comment:
should this always be adding the field to the first frame?

Reply:
I updated the section so intent is more clear